lib/Transforms/ConvertToCiphertextSemantics/ConvertToCiphertextSemantics.cpp

```diff
@@ -0,0 +1,666 @@
+      ContextAwareConversionPatternRewriter &rewriter) const override {
+    rewriter.modifyOpInPlace(op, [&] { setMaterializedAttr(op); });
+    return success();
+  };
+};
+
+// Convert an op generically, marking it as materialized. Lowest priority
+// because it is only meant to handle ops that don't have special
+// materialization rules.
+struct ConvertAnyAddingMaterializedAttr : public ConvertAnyContextAware<> {
+  ConvertAnyAddingMaterializedAttr(
+      const ContextAwareTypeConverter &typeConverter, MLIRContext *context)
+      : ConvertAnyContextAware(typeConverter, context, /*benefit=*/0) {
+    setDebugName("ConvertAnyAddingMaterializedAttr");
+  }
+
+  LogicalResult finalizeOpModification(
+      Operation *op,
+      ContextAwareConversionPatternRewriter &rewriter) const override {
+    rewriter.modifyOpInPlace(op, [&] { setMaterializedAttr(op); });
+    return success();
+  };
+};
+
+class ConvertAssignLayout
+    : public ContextAwareOpConversionPattern<tensor_ext::AssignLayoutOp> {
+ public:
+  using ContextAwareOpConversionPattern<
+      tensor_ext::AssignLayoutOp>::ContextAwareOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor_ext::AssignLayoutOp op, OpAdaptor adaptor,
+      ContextAwareConversionPatternRewriter &rewriter) const final {
+    // This pattern is different because its inputs do not have a layout,
+    // so the type converter doesn't convert their input types. Instead,
+    // the result's layout is defined by this op.
+    //
+    // The input hence needs to be reshaped to fit the output layout, and
+    // by default we assume all values are repeated cyclically according
+    // to the layout.
+    RankedTensorType ciphertextSemanticType =
+        cast<RankedTensorType>(getTypeConverter()->convertType(
+            op.getResult().getType(), op.getLayout()));
+    LLVM_DEBUG(llvm::dbgs() << "Converting AssignLayoutOp to use result type "
+                            << ciphertextSemanticType << "\n");
+    RankedTensorType dataSemanticType = op.getTensor().getType();
+    Value input = adaptor.getTensor();
+
+    // FIXME: support multi-packed ciphertexts
+    if (ciphertextSemanticType.getRank() != 1) {
+      return op.emitError()
+             << "Does not support packing into multiple ciphertexts yet";
+    }
+
+    if (ciphertextSemanticType == input.getType()) {
+      // No conversion needed, the input is already the right size
+      rewriter.replaceOp(op, input);
+      return success();
+    }
+
+    // Otherwise, tile into the ciphertext
+    // FIXME: is there an op that does a cyclic fill? Would be nicer than
+    // a repeated insert_slice...
+    int dataSize = dataSemanticType.getNumElements();
+    int ciphertextSize = ciphertextSemanticType.getShape().back();
+    if (ciphertextSize % dataSize != 0) {
+      return op.emitError() << "Ciphertext size must be a multiple of data "
+                               "size, or else repetition assumption fails!";
```

Comment:
If it's not a multiple, could we pad the data semantic input value with zeros? that would probably be a tensor.fill with zeros and then replicate (insert slice) along offsets that have size (dataSize + padSize) where padSize = (ciphertextSize / numIters) - dataSize

Reply:
Yeah I think zero-padding makes sense for this case. I might leave it for a followup PR though, just for time's sake.